### PR TITLE
C++: Ensure that only one `Function` exists for every function

### PR DESCRIPTION
**Pull request by jketema**
For DCA analysis see the internal issue that links to this PR.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll [hunk 1]
```diff
@@ -122,7 +122,7 @@ private predicate isFunction(Element el) {
  */
 pragma[noopt]
 private predicate callHasNoTarget(@funbindexpr fc) {
-  exists(Function f |
+  exists(@function f |
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll:125]: DIL before:
```
noopt
`ConstantExprs::callHasNoTarget/1#e6e8caa4`(
  /* @funbindexpr */ interned unique entity fc
)
{
  exists(/* Function::Function */ interned entity f |
    funbind(fc, f) and not(`ConstantExprs::isFunction/1#600714be`(f))
  )
}
```
DIL after:
```
noopt
`ConstantExprs::callHasNoTarget/1#e6e8caa4`(
  /* @funbindexpr */ interned unique entity fc
)
{
  exists(/* @function */ interned entity f |
    funbind(fc, f) and not(`ConstantExprs::isFunction/1#600714be`(f))
  )
}
```

File: cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll [hunk 2]
```diff
@@ -110,8 +110,8 @@ private predicate loopConditionAlwaysUponEntry(ControlFlowNode loop, Expr condit
  * should be in this relation.
  */
 pragma[noinline]
-private predicate isFunction(Element el) {
-  el instanceof Function
+private predicate isFunction(@element el) {
+  el instanceof @function
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll:114]: DIL before:
```
noinline
`ConstantExprs::isFunction/1#600714be`(
  /* Element::Element */ interned unique entity el
)
{
  exists(interned dontcare string _, interned dontcare int _1 |
    functions(el, _, _1)
  ) or
  exists(interned dontcare int _ | exprparents(el, _, el))
}
```
DIL after:
```
noinline
`ConstantExprs::isFunction/1#600714be`(/* @element */ interned unique entity el)
{
  exists(interned dontcare string _, interned dontcare int _1 |
    functions(el, _, _1)
  ) or
  exists(interned dontcare int _ | exprparents(el, _, el))
}
```

File: cpp/ql/lib/change-notes/2023-12-22-unique-function.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+---
+category: fix
+---
+* Under certain circumstances a function declaration that is not also a definition could be associated with a `Function` that did not have the definition as a `FunctionDeclarationEntry`. This is now fixed when only ine definition exists, and a unique `Function` will exist that has both the declaration and the definition as a `FunctionDeclarationEntry`.
```
Comment (jketema) [inline cpp/ql/lib/change-notes/2023-12-22-unique-function.md:4]:
```suggestion
* Under certain circumstances a function declaration that is not also a definition could be associated with a `Function` that did not have the definition as a `FunctionDeclarationEntry`. This is now fixed when only one definition exists, and a unique `Function` will exist that has both the declaration and the definition as a `FunctionDeclarationEntry`.
```

**Review by MathiasVP — APPROVED** [review]
LGTM! 🎉

**Pull request merged**